pkg/server/downloader.go

```diff
@@ -0,0 +1,160 @@
+package server
+
+import (
+	"fmt"
+	"net/http"
+	"os"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/longhorn/sparse-tools/sparse"
+	sparserest "github.com/longhorn/sparse-tools/sparse/rest"
+
+	"github.com/longhorn/backing-image-manager/pkg/types"
+	"github.com/longhorn/backing-image-manager/pkg/util"
+)
+
+type DownloaderFactory interface {
+	NewDownloader() Downloader
+}
+
+type Downloader interface {
+	GetSize(url string) (size int64, err error)
+	Pull(url, filepath string, updater util.ProgressUpdater) (written int64, err error)
+	Receive(port string, filePath string, syncFileOps sparserest.SyncFileOperations) error
+	Send(filePath string, address string) error
+	Cancel()
+}
+
+type BackingImageDownloaderFactory struct{}
+
+type BackingImageDownloader struct {
+	*sync.RWMutex
+	cancelFunc context.CancelFunc
+}
+
+func (f *BackingImageDownloaderFactory) NewDownloader() Downloader {
+	return &BackingImageDownloader{
+		&sync.RWMutex{},
+		nil,
+	}
+}
+
+func (d *BackingImageDownloader) GetSize(url string) (written int64, err error) {
+	return util.GetDownloadSize(url)
+}
+
+func (d *BackingImageDownloader) Pull(url, filepath string, updater util.ProgressUpdater) (written int64, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	d.Lock()
+	d.cancelFunc = cancel
+	d.Unlock()
+
+	defer func() {
+		d.Cancel()
+	}()
+	return util.DownloadFile(ctx, cancel, url, filepath, updater)
+}
+
+func (d *BackingImageDownloader) Receive(port string, filePath string, syncFileOps sparserest.SyncFileOperations) error {
+	if err := sparserest.Server(port, filePath, syncFileOps); err != nil && err != http.ErrServerClosed {
```

Comment:
It seems you don't have `thread safe` for pull/download, receive or send, not sure if this is intentend because in the implementation, a new downloader will be created for each backing image so you will not encounter the concurrency. 

It's ok if the downloader is not designed for thread safe, so probably you can remove unnecessary rwmutex. If it's designed for thread for the future, probably you need to make sure some functions thread-safe if need.

Reply:
Ideally, there won't be a receive request during pulling, and vice versa. But deletion will be invoked regardless of pulling or receiving, we need to avoid this race condition. Considering that the cancelling mechanism hasn't been implemented for the receive function, there is no lock right now. Maybe I need to add the lock in advance.

As for the sending function, it shouldn't happen when the backing image is pulling/receiving files. But I will add a check anyway.

Reply:
Thanks for the explanation. After updating, expect we can merge it soon :+1: 

cc @PhanLe1010

Reply:
I have updated this PR. Not sure if we need to wait for the sync cancellation before merging this PR.